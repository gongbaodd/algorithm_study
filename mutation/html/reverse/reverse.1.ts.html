<!DOCTYPE html>
<html><head><title>reverse.1.ts
 - Stryker report</title>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link rel="stylesheet" href="../strykerResources/bootstrap/css/bootstrap.min.css">
<link rel="stylesheet" href="../strykerResources/highlightjs/styles/default.css">
<link rel="stylesheet" href="../strykerResources/stryker/stryker.css"></head>
<body><img class="stryker-image" alt="Stryker" src="../strykerResources/stryker/stryker-80x80.png">
<div class="container-fluid"><div class="row"><div class="col-lg-12"><h1 class="display-4">reverse.1.ts
 
<small class="text-muted">- Stryker report</small></h1></div></div>
<ol class="breadcrumb"><li class="breadcrumb-item"><a href="../index.html">All files</a></li><li class="breadcrumb-item"><a href="index.html">reverse</a></li>
<li class="breadcrumb-item active">reverse.1.ts</li></ol>
<div class="row"><div class="col-lg-12"><div class="row"><div class="col-sm-11"><table class="table table-sm table-hover table-bordered table-no-top"><thead><tr><th style="width: 20%"><div><span>File / Directory</span></div></th>
<th colspan="2"><div><span>Mutation score</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Killed</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Survived</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Timeout</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># No coverage</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Runtime errors</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Transpile errors</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total detected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total undetected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total mutants</span></div></th></tr></thead>
<tbody><tr><td>reverse.1.ts</td>
<td><div class="progress"><div class="progress-bar bg-success" role="progressbar" aria-valuenow="100.00" aria-valuemin="0" aria-valuemax="100" style="width: 100.00%;">100.00
%</div></div></td>
<th class="text-center text-success" style="width: 10%">100.00</th>
<td class="text-center">9</td>
<td class="text-center">0</td>
<td class="text-center">0</td>
<td class="text-center">0</td>
<td class="text-center">7</td>
<td class="text-center">0</td>
<th class="text-center">9</th>
<th class="text-center">0</th>
<th class="text-center">16</th></tr>
</tbody></table></div></div>
<div class="row"><div class="col-lg-7"><div class="row legend"><form class="col-md-12" novalidate="novalidate">

<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="1" type="checkbox">
Killed
 
(9)</label></div>

<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="4" type="checkbox">
RuntimeError
 
(7)</label></div>

<a href="#" class="stryker-collapse-expand-all">Expand all</a></form></div>
<pre><code class="lang-javascript"><span class="bg-null">import { LNode } from &quot;../lib&quot;;

export const reversefn = (_head: LNode): LNode =&gt; </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="0"><span class="badge badge-secondary">0</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="0">{}</span><span class="stryker-original-code" data-mutant="0"><span class="bg-null">{
  const head = _head;

  if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="1"><span class="badge badge-success">1</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="1">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="2"><span class="badge badge-success">2</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="2">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="PrefixUnaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="3"><span class="badge badge-secondary">3</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="3">head.next</span><span class="stryker-original-code" data-mutant="1"><span class="stryker-original-code" data-mutant="2"><span class="stryker-original-code" data-mutant="3"><span class="bg-success-light">!head.next</span></span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="4"><span class="badge badge-success">4</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="4">{}</span><span class="stryker-original-code" data-mutant="4"><span class="bg-success-light">{
    return head;
  }</span></span><span class="bg-null">

  const first = head.next;
  const reverse = (_node: LNode): LNode =&gt; </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="5"><span class="badge badge-secondary">5</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="5">{}</span><span class="stryker-original-code" data-mutant="5"><span class="bg-null">{
    const node = _node;

    if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="6"><span class="badge badge-secondary">6</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="6">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="7"><span class="badge badge-secondary">7</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="7">false</span><span class="stryker-original-code" data-mutant="6"><span class="stryker-original-code" data-mutant="7"><span class="bg-null">node.next </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="8"><span class="badge badge-secondary">8</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="8">!=</span><span class="stryker-original-code" data-mutant="8"><span class="bg-null">==</span></span><span class="bg-null"> null</span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="9"><span class="badge badge-secondary">9</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="9">{}</span><span class="stryker-original-code" data-mutant="9"><span class="bg-null">{
      return node;
    }</span></span><span class="bg-null">

    const last = reverse(node.next);
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="10"><span class="badge badge-success">10</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="10">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="11"><span class="badge badge-success">11</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="11">true</span><span class="stryker-original-code" data-mutant="10"><span class="stryker-original-code" data-mutant="11"><span class="bg-success-light">node.next.next = node</span></span></span><span class="bg-null">; // reverse the next pointer back to self
    </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="12"><span class="badge badge-success">12</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="12">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="13"><span class="badge badge-success">13</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="13">true</span><span class="stryker-original-code" data-mutant="12"><span class="stryker-original-code" data-mutant="13"><span class="bg-success-light">node.next = null</span></span></span><span class="bg-null">;

    return last;
  }</span></span><span class="bg-null">;

  const last = reverse(first as LNode);
  </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="14"><span class="badge badge-success">14</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="14">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="15"><span class="badge badge-success">15</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="15">true</span><span class="stryker-original-code" data-mutant="14"><span class="stryker-original-code" data-mutant="15"><span class="bg-success-light">head.next = last</span></span></span><span class="bg-null">;

  return head;
}</span></span><span class="bg-null">;</span>
</code></pre></div>
<div class="col-lg-5"><table class="table table-sm table-hover mutant-table"><thead><tr><th>#</th>
<th>Mutator</th>
<th>State</th>
<th>Location</th>
<th>Original</th>
<th>Replacement</th></tr></thead>
<tbody><tr><th>0</th>
<td>Block</td>
<td>RuntimeError</td>
<td>2
:
50</td>
<td><code>{
   ...  ;
}</code></td>
<td><code>{}</code></td></tr>
<tr><th>1</th>
<td>IfStatement</td>
<td>Killed</td>
<td>5
:
6</td>
<td><code>!    .</code></td>
<td><code></code></td></tr>
<tr><th>2</th>
<td>IfStatement</td>
<td>Killed</td>
<td>5
:
6</td>
<td><code>!    .</code></td>
<td><code></code></td></tr>
<tr><th>3</th>
<td>PrefixUnaryExpression</td>
<td>RuntimeError</td>
<td>5
:
6</td>
<td><code>!    .</code></td>
<td><code>.</code></td></tr>
<tr><th>4</th>
<td>Block</td>
<td>Killed</td>
<td>5
:
18</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>5</th>
<td>Block</td>
<td>RuntimeError</td>
<td>10
:
43</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>6</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>13
:
8</td>
<td><code>.     ==</code></td>
<td><code></code></td></tr>
<tr><th>7</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>13
:
8</td>
<td><code>.     ==</code></td>
<td><code></code></td></tr>
<tr><th>8</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>13
:
18</td>
<td><code>==</code></td>
<td><code>!=</code></td></tr>
<tr><th>9</th>
<td>Block</td>
<td>RuntimeError</td>
<td>13
:
27</td>
<td><code>{
   ...    }</code></td>
<td><code>{}</code></td></tr>
<tr><th>10</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>18
:
4</td>
<td><code>.    .     =</code></td>
<td><code></code></td></tr>
<tr><th>11</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>18
:
4</td>
<td><code>.    .     =</code></td>
<td><code></code></td></tr>
<tr><th>12</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>19
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>13</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>19
:
4</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>14</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>25
:
2</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>15</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>25
:
2</td>
<td><code>.     =</code></td>
<td><code></code></td></tr></tbody></table></div></div></div></div>
<footer>Generated with stryker-html-reporter generator. Visit the 
<a href="https://stryker-mutator.io" target="_blank">Stryker website</a></footer></div>
<script src="../strykerResources/popper.js/dist/umd/popper.min.js" defer="defer"></script>
<script src="../strykerResources/jquery/dist/jquery.slim.min.js" defer="defer"></script>
<script src="../strykerResources/bootstrap/js/bootstrap.min.js" defer="defer"></script>
<script src="../strykerResources/stryker.js" defer="defer"></script></body></html>