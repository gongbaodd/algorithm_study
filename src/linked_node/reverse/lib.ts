import { LNode, result as lib_result } from "../lib";

const CHAIN_LEN = 8;

export { LNode };

export function init(): LNode {
  let i = 0;

  const head = new LNode(i);
  let tmp: LNode;
  let cur = head;

  while (1) {
    i++;
    tmp = new LNode(i);
    cur.next = tmp;
    cur = tmp;

    if (i >= CHAIN_LEN - 1) {
      break;
    }
  }

  return head;
}

export function result(reverse_fn: (node: LNode) => void) {
  const head = init();
  lib_result(init, reverse_fn);
}
