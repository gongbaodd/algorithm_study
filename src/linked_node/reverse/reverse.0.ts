import { LNode } from "../lib";

export function reversefn(_head: LNode): LNode {
  const head = _head;

  if (head.next == null) {
    return head;
  }

  let cur = head.next;
  let { next } = cur;
  let pre = cur;
  pre.next = null; // in case of a circular reference

  for (;;) {
    cur = next as LNode;
    next = cur.next;

    if (!next) {
      head.next = cur;
      cur.next = pre;
      break;
    }

    cur.next = pre;
    pre = cur;
  }

  return head;
}
