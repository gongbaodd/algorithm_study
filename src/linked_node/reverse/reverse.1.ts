/* eslint-disable no-param-reassign */
import { LNode, result } from "./lib";

result(head => {
  const first = head.next;
  const reverse = (node: LNode): LNode => {
    if (node.next == null) {
      return node;
    }

    const last = reverse(node.next);
    node.next.next = node; // reverse the next pointer back to self
    node.next = null;

    return last;
  };

  const last = reverse(first as LNode);
  head.next = last;
});
