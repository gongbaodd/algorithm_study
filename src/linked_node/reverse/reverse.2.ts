/* eslint-disable no-param-reassign */
import { result, LNode } from "./lib";

result(
  (head): LNode => {
    if (head.next == null) {
      return head;
    }

    const current = head.next;
    let { next } = current;

    for (;;) {
      if (next == null) {
        break;
      }

      const first = head.next;
      current.next = next.next;
      head.next = next;
      next.next = first;
      next = current.next;
    }

    return head;
  }
);
