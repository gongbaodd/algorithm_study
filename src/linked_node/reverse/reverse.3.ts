import { LNode } from "../lib";

export const reverseFn = (_head: LNode): LNode => {
  const head = _head;
  const arr = [];

  let curr = head.next;
  while (curr) {
    const { next } = curr;
    arr.push(curr);
    curr.next = null;
    curr = next;
  }

  for (let i = arr.length; i > 0; i -= 1) {
    arr[i - 1].next = arr[i - 2] ?? null;
  }

  head.next = arr[arr.length - 1];

  return head;
};
