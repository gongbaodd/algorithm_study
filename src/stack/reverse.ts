import { Stack } from "./lib";

function moveBottomToTop(s: Stack<number>) {
  if (s.isEmpty()) {
    return;
  }

  const top1 = s.pop() as number;

  if (s.isEmpty()) {
    s.push(top1);
  } else {
    moveBottomToTop(s);
    const top2 = s.pop() as number;
    s.push(top1);
    s.push(top2);
  }
}

export function reverse(s: Stack<number>) {
  if (s.isEmpty()) {
    return;
  }
  moveBottomToTop(s);
  const top = s.pop() as number;
  reverse(s);
  s.push(top);
}
