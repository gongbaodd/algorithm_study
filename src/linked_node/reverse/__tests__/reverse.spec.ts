import { reversefn as r0 } from "../reverse.0"; // normal
import { reversefn as r1 } from "../reverse.1"; // inverse functions
import { reverseFn as r2 } from "../reverse.2"; // insert
import { reverseFn as r3 } from "../reverse.3";
import { init } from "../../lib";

describe("reverse", () => {
  describe("empty chain", () => {
    it("should return origin chain", () => {
      const head = init([]);
      expect(r0(head)).toMatchObject(head);
      expect(r1(head)).toMatchObject(head);
      expect(r2(head)).toMatchObject(head);
      expect(r3(head)).toMatchObject(head);
    });
  });
  describe("normal chain", () => {
    it("should reverse chain", () => {
      expect(r0(init([1, 2, 3, 4, 5, 6, 7]))).toMatchSnapshot("normal");
      expect(r1(init([1, 2, 3, 4, 5, 6, 7]))).toMatchSnapshot(
        "inverse function"
      );
      expect(r2(init([1, 2, 3, 4, 5, 6, 7]))).toMatchSnapshot("insert");
      expect(r3(init([1, 2, 3, 4, 5, 6, 7]))).toMatchSnapshot("hash set");
    });
  });
});
