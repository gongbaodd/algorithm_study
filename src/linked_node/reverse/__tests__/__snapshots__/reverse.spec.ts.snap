// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reverse normal chain should reverse chain: hash set 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 7,
    "next": LNode {
      "data": 6,
      "next": LNode {
        "data": 5,
        "next": LNode {
          "data": 4,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 2,
              "next": LNode {
                "data": 1,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`reverse normal chain should reverse chain: insert 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 7,
    "next": LNode {
      "data": 6,
      "next": LNode {
        "data": 5,
        "next": LNode {
          "data": 4,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 2,
              "next": LNode {
                "data": 1,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`reverse normal chain should reverse chain: inverse function 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 7,
    "next": LNode {
      "data": 6,
      "next": LNode {
        "data": 5,
        "next": LNode {
          "data": 4,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 2,
              "next": LNode {
                "data": 1,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`reverse normal chain should reverse chain: normal 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 7,
    "next": LNode {
      "data": 6,
      "next": LNode {
        "data": 5,
        "next": LNode {
          "data": 4,
          "next": LNode {
            "data": 3,
            "next": LNode {
              "data": 2,
              "next": LNode {
                "data": 1,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;
